### ISSUE TITLE
Fix pypy3 tests

### ISSUE THREAD

**Issue by BeyondEvil** [pytest-dev/pytest-html#482]
At least two issues to address here:

1. pypy3-mac fails, this might be out of our control but likely related to python 3.6 EOL: https://github.com/actions/setup-python/issues/311
2. pypy3 tests are super slow, this will need some investigation.

**Comment by mattip** [comment]
xref #485. As a PyPy dev, we have dropped support for PyPy3.6, and are going to drop PyPy3.7 soon.

A known issue is that coverage, which on CPython runs as a c-extension and in PyPy runs in pure-python mode, breaks the PyPy JIT so the coverage test run takes 3x-4x more time on PyPy than on CPython.

**Comment by BeyondEvil** [comment]
> xref #485. As a PyPy dev, we have dropped support for PyPy3.6, and are going to drop PyPy3.7 soon.
> 
> A known issue is that coverage, which on CPython runs as a c-extension and in PyPy runs in pure-python mode, breaks the PyPy JIT so the coverage test run takes 3x-4x more time on PyPy than on CPython.

Good to know, thanks!

As you saw in the PR, I'm trying to get it to work on 3.8.

Ah, interesting re. coverage! I have an issue to fix our coverage stuff as well - will keep this in mind then. 👍

**Comment by mattip** [comment]
No, that is indeed strange. We expect tests to run 2x slower since that is roughly the speed of our interpreter vs. CPython (without the JIT) and tests typically do not run long enough for the JIT to kick in. How would you suggest I run the tests directly (without tox) so I can profile?

**Comment by BeyondEvil** [comment]
> No, that is indeed strange. We expect tests to run 2x slower since that is roughly the speed of our interpreter vs. CPython (without the JIT) and tests typically do not run long enough for the JIT to kick in. How would you suggest I run the tests directly (without tox) so I can profile?

Put yourself in a venv with pypy3.8 (that's the version used in GHA) and just run `pytest`. Make sure you install the dependencies listed in `.tox`

**Comment by mattip** [comment]
I did not properly install pytest-html. Now I get a different error, the html does not capture the log:
```
html = '<!DOCTYPE html>\n<html>\n  <head>\n    <meta charset="utf-8"/>\n    <title>report.html</title>\n    <link href="asset...colspan="4">\n            <div class="empty log">No log output captured.</div></td></tr></tbody></table></body></html>'
test_outcome = 'rerun', test_outcome_number = 0, label = None
```

**Comment by mattip** [comment]
On ubuntu 20.04, when I run with PyPy 3.8 I get 20 seconds, with CPython 3.8 10s. This is the 2x I would expect. Running CPython with coverage does not slow down things very much, maybe 11s. On CI, ubuntu20.04, the relationship is again 2x. One windows CI shows a worse ratio: CPython is about 3x slower.

I think this is all within the range of expectations.

**Comment by mattip** [comment]
I missed the 4x on macOS. Now that we see the problem we can try to work on it. Very strange that macOS and windows are that much slower. @cfbloz.

**Comment by mattip** [comment]
Could I add something like [this](https://github.com/napari/napari/pull/4518) to the test setup to upload `--durations` timings as an artifact, and then compare them via https://carreau.github.io/pytest-json-report-viewer/ ?

**Comment by BeyondEvil** [comment]
> Could I add something like [this](https://github.com/napari/napari/pull/4518) to the test setup to upload `--durations` timings as an artifact, and then compare them via https://carreau.github.io/pytest-json-report-viewer/ ?

Absolutely!

**Comment by cfbolz** [comment]
FWIW, I have been running the pytest-html test suite with coverage locally as a kind of real world benchmark for coverage performance (see linked issue). seems we can speed up coverage on pypy for pytest-html (and I suspect many other projects) by a decent amount.
